dex/networks/btc/script.go

```diff
@@ -588,43 +663,45 @@ func InputInfo(pkScript, redeemScript []byte, chainParams *chaincfg.Params) (*Sp
 // contract must be provided for the search algorithm to verify the correct data
 // push. Only contracts of length SwapContractSize that can be validated by
 // ExtractSwapDetails are recognized.
-func FindKeyPush(sigScript, contractHash []byte, chainParams *chaincfg.Params) ([]byte, error) {
-	dataPushes, err := txscript.PushedData(sigScript)
+func FindKeyPush(txIn *wire.TxIn, contractHash []byte, segwit bool, chainParams *chaincfg.Params) ([]byte, error) {
+	var redeemScript, secret []byte
+	var hasher func([]byte) []byte
+	if segwit {
+		if len(txIn.Witness) != 5 {
+			return nil, fmt.Errorf("sigScript should contain 5 data pushes. Found %d", len(txIn.Witness))
+		}
+		secret, redeemScript = txIn.Witness[2], txIn.Witness[4]
+		hasher = func(b []byte) []byte {
+			h := sha256.Sum256(b)
+			return h[:]
+		}
+	} else {
+		pushes, err := txscript.PushedData(txIn.SignatureScript)
+		if err != nil {
+			return nil, fmt.Errorf("sigScript PushedData(%x): %w", txIn.SignatureScript, err)
+		}
+		if len(pushes) != 4 {
+			return nil, fmt.Errorf("sigScript should contain 4 data pushes. Found %d", len(pushes))
+		}
+		secret, redeemScript = pushes[2], pushes[3]
+		hasher = btcutil.Hash160
+	}
+	return findKeyPush(redeemScript, secret, contractHash, chainParams, hasher)
+}
+
+func findKeyPush(redeemScript, secret, contractHash []byte, chainParams *chaincfg.Params, hasher func([]byte) []byte) ([]byte, error) {
+	_, _, _, keyHash, err := ExtractSwapDetails(redeemScript, true, chainParams)
```

Comment:
Worth a comment that segwit arg is irrelevant since addresses aren't used and both addresses are formed from compatible script address data (pubkey hashes).